### ISSUE TITLE
[index layer]./aten/src/ATen/native/cuda/IndexKernel.cu:92: Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed

### ISSUE THREAD

**Issue by peri044** [pytorch/pytorch#116341]
### 🐛 Describe the bug

Error: 
```
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [2,0,0], thread: [32,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [2,0,0], thread: [33,
0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed
```

Steps to reproduce: 
```py
class DDS(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x, mask):
        out = x[mask]
        out = self.relu(out)
        return out

model = DDS().eval().cuda()
x = torch.randn(1, 3, 4, 4).cuda()
y = torch.rand((1, 3, 4, 4), device="cuda") < 0.9
y = y.to(torch.int32)
inputs=[x, y]
ep = torch.export.export(model, tuple(inputs))
gm = ep.module()
from torch.fx.passes.shape_prop import ShapeProp
ShapeProp(gm).propagate(*inputs)
```

### Versions

[pip3] mypy==1.5.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.2
[pip3] onnx==1.14.1
[pip3] onnxruntime==1.16.3
[pip3] pytorch-quantization==2.1.2
[pip3] pytorch-sphinx-theme==0.0.24
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.2.0.dev20231127+cu121
[pip3] torch-tensorrt==2.2.0.dev0+7ad927248
[pip3] torchprofile==0.0.4
[pip3] torchvision==0.17.0.dev20231127+cu121
[pip3] triton==2.1.0

**Comment by angelayi** [comment]
Running the model eagerly seems to produce the same error: 
```python
class DDS(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x, mask):
        out = x[mask]
        out = self.relu(out)
        return out

model = DDS().eval().cuda()
x = torch.randn(1, 3, 4, 4).cuda()
y = torch.rand((1, 3, 4, 4), device="cuda") < 0.9
y = y.to(torch.int32)
inputs=(x, y)
model(*inputs)
```

**Comment by jbschlosser** [comment]
Note that indexing with an int32 dtype tensor is not going to do boolean masking as it appears is expected based on the code snippet. Rather, the first dim of `x` will be indexed based on the values in `y` (i.e. an [`index_select()`](https://pytorch.org/docs/stable/generated/torch.index_select.html) on `dim=0`).

In your example above, the first dim of `x` will randomly be indexed with either 0 or 1, and it only has size 1, so you'll get an error whenever 1 is used as the index.

Closing as expected behavior, but feel free to reopen with more info if you're convinced that a PyTorch bug is present.

**Comment by peri044** [comment]
Thanks @jbschlosser. It sounds my interpretation of `x[mask]` is incorrect. Now the error makes sense. One follow up though: How do you know index_select happens on `dim=0` if we specify `x[mask]` ? Is dim=0 the default ?
